### PR TITLE
Add CLI and SchemaCentralizer

### PR DESCRIPTION
**Pull request by dickermoshe**
2 Changes:

Adds a CLI, pretty simple to use.

Adds a `SchemaCentralizer` which moves all complex schemas to the components section.
Before this, it would generate dynamic for anything not declared in components.
This is only ran when using the CLI.

Addresses Issue https://github.com/tazatechnology/openapi_spec/issues/48

### CODE REVIEW

**Comment by walsha2** [comment]
@dickermoshe good stuff! Let me take a look and mess with this a bit.

Do you have some example specs that make use of these new changes? Can you share the spec (or an abridged version) and the before/after resulting code gen?

**Comment by dickermoshe** [comment]
Seems that I broke it right before pushing it!
I'll ping you when I've got it working again

**Comment by dickermoshe** [comment]
@walsha2 

Here is a full example.

<details><summary>Initial Schema</summary>
<p>

```
openapi: 3.1.0
info:
  title: Swagger Petstore - OpenAPI 3.1
  contact:
    email: apiteam@swagger.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.11
externalDocs:
  description: Find out more about Swagger
  url: http://swagger.io
servers:
  - url: https://petstore3.swagger.io/api/v3
paths:
  /pet:
    put:
      tags:
        - pet
      summary: Update an existing pet
      description: Update an existing pet by Id
      operationId: updatePet
      parameters:
        - name: status
          in: query
          description: Status values that need to be considered for filter
          schema:
            required:
              - name
              - photoUrls
            type: object
            properties:
              id:
                type: integer
                format: int64
                examples: [10]
              name:
                type: string
                examples: [doggie]
              photoUrls:
                type: array
                xml:
                  wrapped: true
                items:
                  type: string
                  xml:
                    name: photoUrl
              status:
                type: string
                description: pet status in the store
                enum:
                  - available
                  - pending
                  - sold
      requestBody:
        description: Update an existent pet in the store
        content:
          application/json:
            schema:
              required:
                - name
                - photoUrls
              type: object
              properties:
                id:
                  type: integer
                  format: int64
                  examples: [10]
                name:
                  type: string
                  examples: [doggie]
                photoUrls:
                  type: array
                  xml:
                    wrapped: true
                  items:
                    type: string
                    xml:
                      name: photoUrl
                status:
                  type: string
                  description: pet status in the store
                  enum:
                    - available
                    - pending
                    - sold

        required: true
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                required:
                  - name
                  - photoUrls
                type: object
                properties:
                  id:
                    type: integer
                    format: int64
                    examples: [10]
                  name:
                    type: string
                    examples: [doggie]
                  photoUrls:
                    type: array
                    xml:
                      wrapped: true
                    items:
                      type: string
                      xml:
                        name: photoUrl
                  status:
                    type: string
                    description: pet status in the store
                    enum:
                      - available
                      - pending
                      - sold
``` 

</p>
</details> 


Before:
```dart
  Future<Map<String, dynamic>> updatePet({  //<< No Type!
    Map<String, dynamic>? status, //<< No Type!
    required Map<String, dynamic> request, // << No Type!
  }) async {
    final r = await makeRequest(
      baseUrl: 'https://petstore3.swagger.io/api/v3',
      path: '/pet',
      method: HttpMethod.put,
      isMultipart: false,
      requestType: 'application/json',
      responseType: 'application/json',
      body: request,
      queryParams: {
        if (status != null) 'status': status,
      },
    );
    return Map<String, dynamic>.from(_jsonDecode(r));
  }

```

After:
```dart
  Future<UpdatePetResponse> updatePet({
    UpdatePetStatus? status,
    required UpdatePetRequestBody request,
  }) async {
    final r = await makeRequest(
      baseUrl: 'https://petstore3.swagger.io/api/v3',
      path: '/pet',
      method: HttpMethod.put,
      isMultipart: false,
      requestType: 'application/json',
      responseType: 'application/json',
      body: request,
      queryParams: {
        if (status != null) 'status': status,
      },
    );
    return UpdatePetResponse.fromJson(_jsonDecode(r));
  }
```

The really cool part of your package is that I can export what the schema looks like after validation.
I converted it to yaml so it's easier to compare.

<details><summary>Exported Schema</summary>
<p>

```
openapi: 3.1.0
info:
  title: Swagger Petstore - OpenAPI 3.1
  contact:
    email: apiteam@swagger.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.11
externalDocs:
  description: Find out more about Swagger
  url: http://swagger.io
servers:
  - url: https://petstore3.swagger.io/api/v3
paths:
  /pet:
    put:
      tags:
        - pet
      summary: Update an existing pet
      description: Update an existing pet by Id
      operationId: updatePet
      parameters:
        - name: status
          in: query
          description: Status values that need to be considered for filter
          schema:
            $ref: '#/components/schemas/UpdatePetStatus'
      requestBody:
        description: Update an existent pet in the store
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/UpdatePetRequestBody'
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UpdatePetResponse'
components:
  schemas:
    UpdatePetRequestBody:
      required:
        - name
        - photoUrls
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        photoUrls:
          type: array
          xml:
            wrapped: true
          items:
            type: string
            xml:
              name: photoUrl
        status:
          type: string
          description: pet status in the store
          enum:
            - available
            - pending
            - sold
    UpdatePetResponse:
      required:
        - name
        - photoUrls
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        photoUrls:
          type: array
          xml:
            wrapped: true
          items:
            type: string
            xml:
              name: photoUrl
        status:
          type: string
          description: pet status in the store
          enum:
            - available
            - pending
            - sold
    UpdatePetStatus:
      required:
        - name
        - photoUrls
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        photoUrls:
          type: array
          xml:
            wrapped: true
          items:
            type: string
            xml:
              name: photoUrl
        status:
          type: string
          description: pet status in the store
          enum:
            - available
            - pending
            - sold
``` 


</p>
</details>

**Comment by dickermoshe** [comment]
I'm down to write test or make any refactors if you want

**Comment by dickermoshe** [comment]
Just curious, why don't you use dio?

**Comment by walsha2** [comment]
> I'm down to write test or make any refactors if you want

Can you rebase using the latest `main` branch? I needed to fix a missed unit testing issue. It is now fixed on `main`. 

Once you rebase, there are a few issues that need to resolved or clarified. See file change comments.

**Comment by walsha2** [comment]
> Just curious, why don't you use dio?

Simple [http](https://pub.dev/packages/http) is a first party package from the Dart core team, `dio` is not. `http` does everything that is required for standard REST API operations, no need to use another third party package.

**Review by walsha2 — CHANGES_REQUESTED** [review]
@dickermoshe a few comments and some uncovered bugs. Let's get these resolved first and then I might have a few more feedback items.

CLI looks good!

File: lib/src/version.dart
Comment (dickermoshe) [reply lib/src/version.dart:2]: You will need to run `build_runner` before each release
This package in pubspec does that
```
  build_verify: ^3.1.0
  build_version: ^2.1.1
```

File: lib/src/commands/generate.dart
Comment (dickermoshe) [reply lib/src/commands/generate.dart:137]: Fixed

File: lib/src/commands/generate.dart
Comment (walsha2) [reply lib/src/commands/generate.dart]: > I respectfully disagree with the approach of not generating data models from the definitions found in the paths

If the API author wanted to create a reusable data model, that would have done so in their API specification and it would show up in the components.

> The reality is that many developers rely on backend frameworks or third-party APIs to generate the OpenAPI spec, rather than writing it from scratch.

That is correct. FastAPI in Python is a great option and there are plenty of examples where request/response data models can/should be specified by the author. Again, the API author can define this. It seems that if it is missing from the spec (as a pre-defined component), it is intentional.

> No other major generator takes the approach you described,

What is the output of the Dart [openapi-generator](https://github.com/OpenAPITools/openapi-generator) for the same spec you provided? Did it provide a typed response?

> I assumed that this would be opt-out.

Due to previous described reasons in other comments, this is difficult to get right because the component is not explicitly defined.

> DevX and popularity will suffer greatly.

"suffer greatly"? That might be a stretch. All I am saying is to add a `centralize=true` flag option. Again, look at the [openapi-generator](https://github.com/OpenAPITools/openapi-generator) options and tell me how that dev experience is. Automatic code generation comes with nuances that there need to be flags in place. There is no once size fits all.

Comment (dickermoshe) [reply lib/src/commands/generate.dart]: Just tried it, it creates a model for the request.

Here is a schema with no components, everything is declared in `paths`.
<details><summary>Schema</summary>
<p>

```
openapi: 3.1.0
info:
  title: Swagger Petstore - OpenAPI 3.1
  contact:
    email: apiteam@swagger.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.11
externalDocs:
  description: Find out more about Swagger
  url: http://swagger.io
servers:
  - url: https://petstore3.swagger.io/api/v3
paths:
  /pet:
    put:
      tags:
        - pet
      summary: Update an existing pet
      description: Update an existing pet by Id
      operationId: updatePet
      parameters:
        - name: status
          in: query
          description: Status values that need to be considered for filter
          schema:
            required:
              - name
              - photoUrls
            type: object
            properties:
              id:
                type: integer
                format: int64
                examples: [10]
              name:
                type: string
                examples: [doggie]
              photoUrls:
                type: array
                xml:
                  wrapped: true
                items:
                  type: string
                  xml:
                    name: photoUrl
              status:
                type: string
                description: pet status in the store
                enum:
                  - available
                  - pending
                  - sold
      requestBody:
        description: Update an existent pet in the store
        content:
          application/json:
            schema:
              required:
                - name
                - photoUrls
              type: object
              properties:
                id:
                  type: integer
                  format: int64
                  examples: [10]
                name:
                  type: string
                  examples: [doggie]
                photoUrls:
                  type: array
                  xml:
                    wrapped: true
                  items:
                    type: string
                    xml:
                      name: photoUrl
                status:
                  type: string
                  description: pet status in the store
                  enum:
                    - available
                    - pending
                    - sold

        required: true
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                required:
                  - name
                  - photoUrls
                type: object
                properties:
                  id:
                    type: integer
                    format: int64
                    examples: [10]
                  name:
                    type: string
                    examples: [doggie]
                  photoUrls:
                    type: array
                    xml:
                      wrapped: true
                    items:
                      type: string
                      xml:
                        name: photoUrl
                  status:
                    type: string
                    description: pet status in the store
                    enum:
                      - available
                      - pending
                      - sold
``` 

</p>
</details> 


It generates a model for it `update_pet_request.dart`
```dart
//
// AUTO-GENERATED FILE, DO NOT MODIFY!
//

// ignore_for_file: unused_element
import 'package:built_collection/built_collection.dart';
import 'package:built_value/built_value.dart';
import 'package:built_value/serializer.dart';

part 'update_pet_request.g.dart';

/// UpdatePetRequest
///
/// Properties:
/// * [id] 
/// * [name] 
/// * [photoUrls] 
/// * [status] - pet status in the store
@BuiltValue()
abstract class UpdatePetRequest implements Built<UpdatePetRequest, UpdatePetRequestBuilder> {
  @BuiltValueField(wireName: r'id')
  int? get id;

  @BuiltValueField(wireName: r'name')
  String get name;

  @BuiltValueField(wireName: r'photoUrls')
  BuiltList<String> get photoUrls;

  /// pet status in the store
  @BuiltValueField(wireName: r'status')
  UpdatePetRequestStatusEnum? get status;
  // enum statusEnum {  available,  pending,  sold,  };

  UpdatePetRequest._();

  factory UpdatePetRequest([void updates(UpdatePetRequestBuilder b)]) = _$UpdatePetRequest;

  @BuiltValueHook(initializeBuilder: true)
  static void _defaults(UpdatePetRequestBuilder b) => b;

  @BuiltValueSerializer(custom: true)
  static Serializer<UpdatePetRequest> get serializer => _$UpdatePetRequestSerializer();
}

class _$UpdatePetRequestSerializer implements PrimitiveSerializer<UpdatePetRequest> {
  @override
  final Iterable<Type> types = const [UpdatePetRequest, _$UpdatePetRequest];

  @override
  final String wireName = r'UpdatePetRequest';

  Iterable<Object?> _serializeProperties(
    Serializers serializers,
    UpdatePetRequest object, {
    FullType specifiedType = FullType.unspecified,
  }) sync* {
    if (object.id != null) {
      yield r'id';
      yield serializers.serialize(
        object.id,
        specifiedType: const FullType(int),
      );
    }
    yield r'name';
    yield serializers.serialize(
      object.name,
      specifiedType: const FullType(String),
    );
    yield r'photoUrls';
    yield serializers.serialize(
      object.photoUrls,
      specifiedType: const FullType(BuiltList, [FullType(String)]),
    );
    if (object.status != null) {
      yield r'status';
      yield serializers.serialize(
        object.status,
        specifiedType: const FullType(UpdatePetRequestStatusEnum),
      );
    }
  }

  @override
  Object serialize(
    Serializers serializers,
    UpdatePetRequest object, {
    FullType specifiedType = FullType.unspecified,
  }) {
    return _serializeProperties(serializers, object, specifiedType: specifiedType).toList();
  }

  void _deserializeProperties(
    Serializers serializers,
    Object serialized, {
    FullType specifiedType = FullType.unspecified,
    required List<Object?> serializedList,
    required UpdatePetRequestBuilder result,
    required List<Object?> unhandled,
  }) {
    for (var i = 0; i < serializedList.length; i += 2) {
      final key = serializedList[i] as String;
      final value = serializedList[i + 1];
      switch (key) {
        case r'id':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(int),
          ) as int;
          result.id = valueDes;
          break;
        case r'name':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(String),
          ) as String;
          result.name = valueDes;
          break;
        case r'photoUrls':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(BuiltList, [FullType(String)]),
          ) as BuiltList<String>;
          result.photoUrls.replace(valueDes);
          break;
        case r'status':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(UpdatePetRequestStatusEnum),
          ) as UpdatePetRequestStatusEnum;
          result.status = valueDes;
          break;
        default:
          unhandled.add(key);
          unhandled.add(value);
          break;
      }
    }
  }

  @override
  UpdatePetRequest deserialize(
    Serializers serializers,
    Object serialized, {
    FullType specifiedType = FullType.unspecified,
  }) {
    final result = UpdatePetRequestBuilder();
    final serializedList = (serialized as Iterable<Object?>).toList();
    final unhandled = <Object?>[];
    _deserializeProperties(
      serializers,
      serialized,
      specifiedType: specifiedType,
      serializedList: serializedList,
      unhandled: unhandled,
      result: result,
    );
    return result.build();
  }
}

class UpdatePetRequestStatusEnum extends EnumClass {

  /// pet status in the store
  @BuiltValueEnumConst(wireName: r'available')
  static const UpdatePetRequestStatusEnum available = _$updatePetRequestStatusEnum_available;
  /// pet status in the store
  @BuiltValueEnumConst(wireName: r'pending')
  static const UpdatePetRequestStatusEnum pending = _$updatePetRequestStatusEnum_pending;
  /// pet status in the store
  @BuiltValueEnumConst(wireName: r'sold')
  static const UpdatePetRequestStatusEnum sold = _$updatePetRequestStatusEnum_sold;

  static Serializer<UpdatePetRequestStatusEnum> get serializer => _$updatePetRequestStatusEnumSerializer;

  const UpdatePetRequestStatusEnum._(String name): super(name);

  static BuiltSet<UpdatePetRequestStatusEnum> get values => _$updatePetRequestStatusEnumValues;
  static UpdatePetRequestStatusEnum valueOf(String name) => _$updatePetRequestStatusEnumValueOf(name);
}


```

Comment (dickermoshe) [reply lib/src/commands/generate.dart]: > All I am saying is to add a centralize=true flag option.  

This doesn't mean anything to someone who doesnt understand the OpenAPI spec, which is most people.  
I name the flag `just-work-correctly`, in which case it would be kinda funny if it wasn't on by default. 😂

Comment (walsha2) [reply lib/src/commands/generate.dart]: Man... that is ugly generated code.

> Just tried it, it creates a model for the request.

Ok cool. I am not completely opposed to it being opt-in or opt-out, it just should be optional. The default can go either way.

Can you rebase your branch on the latest `main`?

Comment (dickermoshe) [reply lib/src/commands/generate.dart]: I won't die on this hill, but as long as it documented very clearly in the docs it's ok I guess.

**To Generate Models for All Objects, use `--centralize`**

Comment (walsha2) [reply lib/src/commands/generate.dart]: > This doesn't mean anything to someone who doesnt understand the OpenAPI spec, which is most people.
I name the flag just-work-correctly, in which case it would be kinda funny if it wasn't on by default.

Facts. You are starting to sell me @dickermoshe

Comment (dickermoshe) [reply lib/src/commands/generate.dart]: You did 99% of the work on this package, It's your call

Comment (dickermoshe) [reply lib/src/commands/generate.dart]: Rebased.

**Comment by dickermoshe** [comment]
I pulled all of the openapi specs from the `openapi-generator` package and ran the generator on all of them found some small bugs, I'll file issues later.

**Comment by walsha2** [comment]
> found some small bugs, I'll file issues later.

Bugs related to the centralize step or to the existing generator? FWIW all of the sample openapi specs are in the `test/oas_examples/` directory and are tested.

**Comment by dickermoshe** [comment]
The existing code, not the centralizer.

**Comment by walsha2** [comment]
> The existing code, not the centralizer.

Great! Yes, please file the issues (sooner the better) and I can take a look. Would prefer that those are fixed before next release.

**Pull request merged**
